File: 2024/9/main.clj
(require '[clojure.string :as str])

(def data (slurp (first *command-line-args*)))

(def lines (str/split data #"\n"))

(def disk-map (mapv #(Character/digit % 10) (first lines)))

(defn rpt-vec [v n] (vec (repeat n v)))

(defn checksum [disk]
  (reduce + (map-indexed (fn [i val] (if (>= val 0) (* i val) 0)) disk)))

(let [[disk files frees files2 frees2] (loop [disk []
                                              cur 0
                                              is-file true
                                              files []
                                              frees []
                                              files2 []
                                              frees2 []
                                              disk-map disk-map]
                                         (if (empty? disk-map)
                                           [disk files frees files2 frees2]
                                           (if is-file
                                             (let [length (first disk-map)
                                                   files' (concat files (vec (range (count disk) (+ (count disk) length))))
                                                   files2' (conj files2 [(count disk) length])
                                                   disk' (concat disk (rpt-vec cur length))]
                                               (recur disk' (inc cur) (not is-file) files' frees files2' frees2 (rest disk-map)))
                                             (let [length (first disk-map)
                                                   frees' (concat frees (vec (range (count disk) (+ (count disk) length))))
                                                   frees2' (conj frees2 [(count disk) length])
                                                   disk' (concat disk (rpt-vec -1 length))]
                                               (recur disk' cur (not is-file) files frees' files2 frees2' (rest disk-map))))))]

  (def disk1 (loop [disk (vec disk)
                    frees frees
                    files files]
               (if (> (first frees) (last files))
                 disk
                 (let [file-idx (last files)
                       free-idx (first frees)
                       disk' (assoc disk
                                    file-idx (disk free-idx)
                                    free-idx (disk file-idx))]
                   (recur disk' (rest frees) (butlast files))))))

  (def disk2 (first (reduce
                     (fn [[disk frees2] [file-idx file-len]]
                       (reduce
                        (fn [[disk frees2 broke] [i [free-idx free-len]]]
                          (if (and (<= file-len free-len) (> file-idx free-idx) (not broke))
                            (let [disk' (reduce
                                         (fn [disk j]
                                           (assoc (vec disk)
                                                  (+ free-idx j) ((vec disk) (+ file-idx j))
                                                  (+ file-idx j) ((vec disk) (+ free-idx j))))
                                         disk
                                         (range file-len))
                                  frees2' (assoc frees2 i [(+ free-idx file-len) (- free-len file-len)])]
                              [disk' frees2' true])
                            [disk frees2 broke]))
                        [disk frees2 false]
                        (map-indexed vector frees2)))
                     [disk frees2]
                     (reverse files2)))))

(println (checksum disk1))
(println (checksum disk2))

File: 2024/9/main.py
import sys

data = open(sys.argv[1]).read().strip()
lines = data.split('\n')
disk_map = [int(n) for n in lines[0]]


def checksum(disk):
    return sum([i*val for i, val in enumerate(disk) if val >= 0])


disk = []
cur = 0
is_file = True
files, frees = [], []
files2, frees2 = [], []
for length in disk_map:
    if is_file:
        files.extend(i for i in range(len(disk), len(disk)+length))
        files2.append((len(disk), length))
        disk.extend([cur] * length)
        cur += 1
    else:
        frees.extend(i for i in range(len(disk), len(disk)+length))
        frees2.append((len(disk), length))
        disk.extend([-1] * length)
    is_file = not is_file

disk2 = disk[::]

while frees[0] <= files[-1]:
    file_idx, free_idx = files[-1], frees[0]
    disk[file_idx], disk[free_idx] = disk[free_idx], disk[file_idx]
    files.pop()
    frees.pop(0)

for file_idx, file_len in files2[::-1]:
    for i, (free_idx, free_len) in enumerate(frees2):
        if file_len <= free_len and file_idx > free_idx:
            for j in range(file_len):
                disk2[free_idx+j], disk2[file_idx+j] = disk2[file_idx+j], disk2[free_idx+j]
            frees2[i] = (free_idx+file_len, free_len-file_len)
            break

print(checksum(disk))
print(checksum(disk2))
